File: Semester 5/PPD/lab7CumTrebe/MPIparalele/MPIparalele/BigIntMpi.cs
using System;
using System.Diagnostics;
using System.Text;
using MPI;

namespace MPIparalele
{
    // Big integer represented as array of digits (base 10)
    // digits[0] is the least significant digit
    public class BigInt
    {
        public int[] Digits { get; set; }
        public int Size { get; set; }
        public int Capacity { get; set; }

        // Initialize a BigInt
        public BigInt(int capacity)
        {
            Digits = new int[capacity];
            Size = 0;
            Capacity = capacity;
        }

        // Create BigInt from string
        public static BigInt FromString(string str)
        {
            int len = str.Length;
            BigInt bi = new BigInt(len + 10);
            bi.Size = len;

            // Reverse the string to store least significant digit first
            for (int i = 0; i < len; i++)
            {
                bi.Digits[i] = str[len - 1 - i] - '0';
            }

            return bi;
        }

        // Create BigInt from integer
        public static BigInt FromInt(int value)
        {
            BigInt bi = new BigInt(20);

            if (value == 0)
            {
                bi.Size = 1;
                bi.Digits[0] = 0;
                return bi;
            }

            int idx = 0;
            while (value > 0)
            {
                bi.Digits[idx++] = value % 10;
                value /= 10;
            }
            bi.Size = idx;

            return bi;
        }

        // Convert BigInt to string
        public string ToString()
        {
            StringBuilder sb = new StringBuilder(Size);

            for (int i = 0; i < Size; i++)
            {
                sb.Append((char)(Digits[Size - 1 - i] + '0'));
            }

            return sb.ToString();
        }

        // Normalize BigInt (remove leading zeros, handle carries)
        public void Normalize()
        {
            // Handle carries
            for (int i = 0; i < Capacity - 1; i++)
            {
                if (Digits[i] >= 10)
                {
                    Digits[i + 1] += Digits[i] / 10;
                    Digits[i] %= 10;
                }
            }

            // Update size
            int newSize = Capacity;
            while (newSize > 1 && Digits[newSize - 1] == 0)
            {
                newSize--;
            }
            Size = newSize;
        }
    }

    public class BigIntMpi
    {
        // Regular O(n²) multiplication using MPI
        public static void BigIntMultiplyMpi(BigInt a, BigInt b, BigInt result, int rank, int size)
        {
            int resultCapacity = a.Size + b.Size + 10;

            // Initialize result
            Array.Clear(result.Digits, 0, result.Capacity);
            result.Size = 0;

            // Distribute work: each process multiplies a chunk of a with all of b
            int chunkSize = (a.Size + size - 1) / size;
            int startIdx = rank * chunkSize;
            int endIdx = (startIdx + chunkSize < a.Size) ? startIdx + chunkSize : a.Size;

            // Each process computes partial result
            for (int i = startIdx; i < endIdx; i++)
            {
                for (int j = 0; j < b.Size; j++)
                {
                    result.Digits[i + j] += a.Digits[i] * b.Digits[j];
                }
            }
        }

        // Regular sequential multiplication (for comparison)
        public static void BigIntMultiplySeq(BigInt a, BigInt b, BigInt result)
        {
            int resultCapacity = a.Size + b.Size + 10;

            Array.Clear(result.Digits, 0, result.Capacity);
            result.Size = 0;

            for (int i = 0; i < a.Size; i++)
            {
                for (int j = 0; j < b.Size; j++)
                {
                    result.Digits[i + j] += a.Digits[i] * b.Digits[j];
                }
            }

            result.Normalize();
        }

        // Karatsuba algorithm for big integers
        public static void BigIntKaratsubaMpi(BigInt a, BigInt b, BigInt result, int rank, int size, Intracommunicator comm)
        {
            // Base case: use regular multiplication for small numbers
            if (a.Size <= 16 || b.Size <= 16)
            {
                BigIntMultiplySeq(a, b, result);
                result.Normalize();
                return;
            }

            // Make both numbers same length
            int maxLen = (a.Size > b.Size) ? a.Size : b.Size;

            // Pad to even length
            if (maxLen % 2 == 1)
            {
                maxLen++;
            }

            // Split at midpoint
            int mid = maxLen / 2;

            // Create split BigInts
            BigInt aLow = new BigInt(mid + 10);
            BigInt aHigh = new BigInt(mid + 10);
            BigInt bLow = new BigInt(mid + 10);
            BigInt bHigh = new BigInt(mid + 10);

            // Copy digits
            int aLowSize = (mid < a.Size) ? mid : a.Size;
            int aHighSize = (a.Size > mid) ? a.Size - mid : 0;
            int bLowSize = (mid < b.Size) ? mid : b.Size;
            int bHighSize = (b.Size > mid) ? b.Size - mid : 0;

            Array.Copy(a.Digits, 0, aLow.Digits, 0, aLowSize);
            aLow.Size = aLowSize;
            if (aHighSize > 0)
            {
                Array.Copy(a.Digits, mid, aHigh.Digits, 0, aHighSize);
                aHigh.Size = aHighSize;
            }
            else
            {
                aHigh.Size = 1;
                aHigh.Digits[0] = 0;
            }

            Array.Copy(b.Digits, 0, bLow.Digits, 0, bLowSize);
            bLow.Size = bLowSize;
            if (bHighSize > 0)
            {
                Array.Copy(b.Digits, mid, bHigh.Digits, 0, bHighSize);
                bHigh.Size = bHighSize;
            }
            else
            {
                bHigh.Size = 1;
                bHigh.Digits[0] = 0;
            }

            // Allocate space for z0, z1, z2
            int zSize = 2 * mid + 10;
            BigInt z0 = new BigInt(zSize);
            BigInt z2 = new BigInt(zSize);

            // Distribute work: rank 0 does z0, rank 1 does z2
            if (rank == 0)
            {
                BigIntKaratsubaMpi(aLow, bLow, z0, rank, size, comm);
                // If only one process, also compute z2
                if (size == 1)
                {
                    BigIntKaratsubaMpi(aHigh, bHigh, z2, rank, size, comm);
                }
            }
            else if (rank == 1 && size > 1)
            {
                BigIntKaratsubaMpi(aHigh, bHigh, z2, rank, size, comm);
            }

            // Compute z1 at rank 0
            BigInt z1 = null;
            if (rank == 0)
            {
                // (a_low + a_high) * (b_low + b_high)
                BigInt aSum = new BigInt(mid + 10);
                BigInt bSum = new BigInt(mid + 10);

                // Add a_low + a_high
                int maxA = (aLow.Size > aHigh.Size) ? aLow.Size : aHigh.Size;
                for (int i = 0; i < maxA; i++)
                {
                    int val = 0;
                    if (i < aLow.Size) val += aLow.Digits[i];
                    if (i < aHigh.Size) val += aHigh.Digits[i];
                    aSum.Digits[i] = val;
                }
                aSum.Size = maxA;
                aSum.Normalize();

                // Add b_low + b_high
                int maxB = (bLow.Size > bHigh.Size) ? bLow.Size : bHigh.Size;
                for (int i = 0; i < maxB; i++)
                {
                    int val = 0;
                    if (i < bLow.Size) val += bLow.Digits[i];
                    if (i < bHigh.Size) val += bHigh.Digits[i];
                    bSum.Digits[i] = val;
                }
                bSum.Size = maxB;
                bSum.Normalize();

                BigInt z1Sum = new BigInt(zSize);
                BigIntKaratsubaMpi(aSum, bSum, z1Sum, rank, size, comm);

                // z1 = z1_sum - z0 - z2
                z1 = new BigInt(zSize);
                Array.Copy(z1Sum.Digits, z1.Digits, z1Sum.Size);
                z1.Size = z1Sum.Size;

                // Subtract z0
                for (int i = 0; i < z0.Size; i++)
                {
                    z1.Digits[i] -= z0.Digits[i];
                }

                // Subtract z2
                for (int i = 0; i < z2.Size; i++)
                {
                    z1.Digits[i] -= z2.Digits[i];
                }

                z1.Normalize();
            }

            // Gather z2 from rank 1
            if (size > 1)
            {
                if (rank == 1)
                {
                    comm.Send(z2.Size, 0, 0);
                    int[] z2Digits = new int[z2.Size];
                    Array.Copy(z2.Digits, z2Digits, z2.Size);
                    comm.Send(z2Digits, 0, 1);
                }
                else if (rank == 0)
                {
                    int z2Size;
                    comm.Receive(1, 0, out z2Size);
                    if (z2Size > z2.Capacity)
                    {
                        int[] newDigits = new int[z2Size];
                        Array.Copy(z2.Digits, newDigits, Math.Min(z2.Size, z2Size));
                        z2.Digits = newDigits;
                        z2.Capacity = z2Size;
                    }
                    int[] z2Digits = new int[z2Size];
                    comm.Receive(1, 1, out z2Digits);
                    Array.Copy(z2Digits, z2.Digits, z2Size);
                    z2.Size = z2Size;
                }
            }

            // Combine results at rank 0: result = z0 + z1*10^mid + z2*10^(2*mid)
            if (rank == 0)
            {
                Array.Clear(result.Digits, 0, result.Capacity);
                result.Size = 0;

                // Add z0
                for (int i = 0; i < z0.Size; i++)
                {
                    result.Digits[i] += z0.Digits[i];
                }

                // Add z1 shifted by mid
                if (z1 != null)
                {
                    for (int i = 0; i < z1.Size; i++)
                    {
                        result.Digits[i + mid] += z1.Digits[i];
                    }
                }

                // Add z2 shifted by 2*mid
                for (int i = 0; i < z2.Size; i++)
                {
                    result.Digits[i + 2 * mid] += z2.Digits[i];
                }

                result.Normalize();
            }
        }

        public static void Main(string[] args)
        {
            using (new MPI.Environment(ref args))
            {
                Intracommunicator comm = Communicator.world;
                int rank = comm.Rank;
                int size = comm.Size;

                if (args.Length < 2)
                {
                    if (rank == 0)
                    {
                        Console.WriteLine($"Usage: {System.Reflection.Assembly.GetExecutingAssembly().GetName().Name} <num1> <num2> [algorithm]");
                        Console.WriteLine("  num1, num2: big integers as strings");
                        Console.WriteLine("  algorithm: 'regular' or 'karatsuba' (default: regular)");
                    }
                    return;
                }

                string num1Str = args[0];
                string num2Str = args[1];
                string algorithm = (args.Length > 2) ? args[2] : "regular";

                BigInt a = null;
                BigInt b = null;
                BigInt result = null;

                if (rank == 0)
                {
                    a = BigInt.FromString(num1Str);
                    b = BigInt.FromString(num2Str);

                    Console.WriteLine($"Number 1: {num1Str}");
                    Console.WriteLine($"Number 2: {num2Str}");
                }

                // Broadcast sizes
                int aSize = 0, bSize = 0;
                if (rank == 0)
                {
                    aSize = a.Size;
                    bSize = b.Size;
                }
                comm.Broadcast(ref aSize, 0);
                comm.Broadcast(ref bSize, 0);

                // Allocate and broadcast BigInts
                if (rank != 0)
                {
                    a = new BigInt(aSize + 10);
                    b = new BigInt(bSize + 10);
                    a.Size = aSize;
                    b.Size = bSize;
                }

                int[] aDigits = new int[aSize];
                int[] bDigits = new int[bSize];
                if (rank == 0)
                {
                    Array.Copy(a.Digits, aDigits, aSize);
                    Array.Copy(b.Digits, bDigits, bSize);
                }
                comm.Broadcast(ref aDigits, 0);
                comm.Broadcast(ref bDigits, 0);
                if (rank != 0)
                {
                    Array.Copy(aDigits, a.Digits, aSize);
                    Array.Copy(bDigits, b.Digits, bSize);
                }

                int resultCapacity = aSize + bSize + 100;
                result = new BigInt(resultCapacity);

                // Time the computation
                Stopwatch sw = Stopwatch.StartNew();

                if (algorithm == "karatsuba")
                {
                    if (rank == 0)
                    {
                        Console.WriteLine($"Using Karatsuba algorithm with {size} processes");
                    }
                    BigIntKaratsubaMpi(a, b, result, rank, size, comm);
                }
                else
                {
                    if (rank == 0)
                    {
                        Console.WriteLine($"Using Regular O(n²) algorithm with {size} processes");
                    }

                    BigInt partialResult = new BigInt(resultCapacity);
                    BigIntMultiplyMpi(a, b, partialResult, rank, size);

                    // Reduce partial results
                    int[] partialDigits = new int[resultCapacity];
                    Array.Copy(partialResult.Digits, partialDigits, resultCapacity);
                    int[] resultDigits = new int[resultCapacity];
                    int root = 0;
                    comm.Reduce(partialDigits, Operation<int>.Add, resultDigits, ref root);

                    if (rank == 0)
                    {
                        Array.Copy(resultDigits, result.Digits, resultCapacity);
                        result.Normalize();
                    }
                }

                sw.Stop();
                double elapsed = sw.Elapsed.TotalSeconds;

                if (rank == 0)
                {
                    string resultStr = result.ToString();
                    Console.WriteLine($"\nResult: {resultStr}");
                    Console.WriteLine($"Time: {elapsed:F6} seconds");
                }
            }
        }
    }
}

File: Semester 5/PPD/lab7/bigint_mpi.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// Big integer represented as array of digits (base 10)
// digits[0] is the least significant digit

typedef struct {
    int *digits;
    int size;
    int capacity;
} BigInt;

// Initialize a BigInt
BigInt* bigint_init(int capacity) {
    BigInt *bi = (BigInt *)malloc(sizeof(BigInt));
    bi->digits = (int *)calloc(capacity, sizeof(int));
    bi->size = 0;
    bi->capacity = capacity;
    return bi;
}

// Free a BigInt
void bigint_free(BigInt *bi) {
    if (bi) {
        free(bi->digits);
        free(bi);
    }
}

// Create BigInt from string
BigInt* bigint_from_string(const char *str) {
    int len = strlen(str);
    BigInt *bi = bigint_init(len + 10);
    bi->size = len;
    
    // Reverse the string to store least significant digit first
    for (int i = 0; i < len; i++) {
        bi->digits[i] = str[len - 1 - i] - '0';
    }
    
    return bi;
}

// Create BigInt from integer
BigInt* bigint_from_int(int value) {
    BigInt *bi = bigint_init(20);
    
    if (value == 0) {
        bi->size = 1;
        bi->digits[0] = 0;
        return bi;
    }
    
    int idx = 0;
    while (value > 0) {
        bi->digits[idx++] = value % 10;
        value /= 10;
    }
    bi->size = idx;
    
    return bi;
}

// Convert BigInt to string
char* bigint_to_string(BigInt *bi) {
    char *str = (char *)malloc((bi->size + 1) * sizeof(char));
    
    for (int i = 0; i < bi->size; i++) {
        str[i] = bi->digits[bi->size - 1 - i] + '0';
    }
    str[bi->size] = '\0';
    
    return str;
}

// Normalize BigInt (remove leading zeros, handle carries)
void bigint_normalize(BigInt *bi) {
    // Handle carries
    for (int i = 0; i < bi->capacity - 1; i++) {
        if (bi->digits[i] >= 10) {
            bi->digits[i + 1] += bi->digits[i] / 10;
            bi->digits[i] %= 10;
        }
    }
    
    // Update size
    int new_size = bi->capacity;
    while (new_size > 1 && bi->digits[new_size - 1] == 0) {
        new_size--;
    }
    bi->size = new_size;
}

// Regular O(n²) multiplication using MPI
void bigint_multiply_mpi(BigInt *a, BigInt *b, BigInt *result, int rank, int size) {
    int result_capacity = a->size + b->size + 10;
    
    // Initialize result
    memset(result->digits, 0, result->capacity * sizeof(int));
    result->size = 0;
    
    // Distribute work: each process multiplies a chunk of a with all of b
    int chunk_size = (a->size + size - 1) / size;
    int start_idx = rank * chunk_size;
    int end_idx = (start_idx + chunk_size < a->size) ? start_idx + chunk_size : a->size;
    
    // Each process computes partial result
    for (int i = start_idx; i < end_idx; i++) {
        for (int j = 0; j < b->size; j++) {
            result->digits[i + j] += a->digits[i] * b->digits[j];
        }
    }
}

// Regular sequential multiplication (for comparison)
void bigint_multiply_seq(BigInt *a, BigInt *b, BigInt *result) {
    int result_capacity = a->size + b->size + 10;
    
    memset(result->digits, 0, result->capacity * sizeof(int));
    result->size = 0;
    
    for (int i = 0; i < a->size; i++) {
        for (int j = 0; j < b->size; j++) {
            result->digits[i + j] += a->digits[i] * b->digits[j];
        }
    }
    
    bigint_normalize(result);
}

// Karatsuba algorithm for big integers
void bigint_karatsuba_mpi(BigInt *a, BigInt *b, BigInt *result, int rank, int size) {
    // Base case: use regular multiplication for small numbers
    if (a->size <= 16 || b->size <= 16) {
        bigint_multiply_seq(a, b, result);
        bigint_normalize(result);
        return;
    }
    
    // Make both numbers same length
    int max_len = (a->size > b->size) ? a->size : b->size;
    
    // Pad to even length
    if (max_len % 2 == 1) {
        max_len++;
    }
    
    // Split at midpoint
    int mid = max_len / 2;
    
    // Create split BigInts
    BigInt *a_low = bigint_init(mid + 10);
    BigInt *a_high = bigint_init(mid + 10);
    BigInt *b_low = bigint_init(mid + 10);
    BigInt *b_high = bigint_init(mid + 10);
    
    // Copy digits
    int a_low_size = (mid < a->size) ? mid : a->size;
    int a_high_size = (a->size > mid) ? a->size - mid : 0;
    int b_low_size = (mid < b->size) ? mid : b->size;
    int b_high_size = (b->size > mid) ? b->size - mid : 0;
    
    memcpy(a_low->digits, a->digits, a_low_size * sizeof(int));
    a_low->size = a_low_size;
    if (a_high_size > 0) {
        memcpy(a_high->digits, a->digits + mid, a_high_size * sizeof(int));
        a_high->size = a_high_size;
    } else {
        a_high->size = 1;
        a_high->digits[0] = 0;
    }
    
    memcpy(b_low->digits, b->digits, b_low_size * sizeof(int));
    b_low->size = b_low_size;
    if (b_high_size > 0) {
        memcpy(b_high->digits, b->digits + mid, b_high_size * sizeof(int));
        b_high->size = b_high_size;
    } else {
        b_high->size = 1;
        b_high->digits[0] = 0;
    }
    
    // Allocate space for z0, z1, z2
    int z_size = 2 * mid + 10;
    BigInt *z0 = bigint_init(z_size);
    BigInt *z2 = bigint_init(z_size);
    
    // Distribute work: rank 0 does z0, rank 1 does z2
    if (rank == 0) {
        bigint_karatsuba_mpi(a_low, b_low, z0, rank, size);
        // If only one process, also compute z2
        if (size == 1) {
            bigint_karatsuba_mpi(a_high, b_high, z2, rank, size);
        }
    } else if (rank == 1 && size > 1) {
        bigint_karatsuba_mpi(a_high, b_high, z2, rank, size);
    }
    
    // Compute z1 at rank 0
    BigInt *z1 = NULL;
    if (rank == 0) {
        // (a_low + a_high) * (b_low + b_high)
        BigInt *a_sum = bigint_init(mid + 10);
        BigInt *b_sum = bigint_init(mid + 10);
        
        // Add a_low + a_high
        int max_a = (a_low->size > a_high->size) ? a_low->size : a_high->size;
        for (int i = 0; i < max_a; i++) {
            int val = 0;
            if (i < a_low->size) val += a_low->digits[i];
            if (i < a_high->size) val += a_high->digits[i];
            a_sum->digits[i] = val;
        }
        a_sum->size = max_a;
        bigint_normalize(a_sum);
        
        // Add b_low + b_high
        int max_b = (b_low->size > b_high->size) ? b_low->size : b_high->size;
        for (int i = 0; i < max_b; i++) {
            int val = 0;
            if (i < b_low->size) val += b_low->digits[i];
            if (i < b_high->size) val += b_high->digits[i];
            b_sum->digits[i] = val;
        }
        b_sum->size = max_b;
        bigint_normalize(b_sum);
        
        BigInt *z1_sum = bigint_init(z_size);
        bigint_karatsuba_mpi(a_sum, b_sum, z1_sum, rank, size);
        
        // z1 = z1_sum - z0 - z2
        z1 = bigint_init(z_size);
        memcpy(z1->digits, z1_sum->digits, z1_sum->size * sizeof(int));
        z1->size = z1_sum->size;
        
        // Subtract z0
        for (int i = 0; i < z0->size; i++) {
            z1->digits[i] -= z0->digits[i];
        }
        
        // Subtract z2
        for (int i = 0; i < z2->size; i++) {
            z1->digits[i] -= z2->digits[i];
        }
        
        bigint_normalize(z1);
        
        bigint_free(a_sum);
        bigint_free(b_sum);
        bigint_free(z1_sum);
    }
    
    // Gather z2 from rank 1
    if (size > 1) {
        if (rank == 1) {
            MPI_Send(&z2->size, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
            MPI_Send(z2->digits, z2->size, MPI_INT, 0, 1, MPI_COMM_WORLD);
        } else if (rank == 0) {
            MPI_Recv(&z2->size, 1, MPI_INT, 1, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
            if (z2->size > z2->capacity) {
                z2->digits = (int *)realloc(z2->digits, z2->size * sizeof(int));
                z2->capacity = z2->size;
            }
            MPI_Recv(z2->digits, z2->size, MPI_INT, 1, 1, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        }
    }
    
    // Combine results at rank 0: result = z0 + z1*10^mid + z2*10^(2*mid)
    if (rank == 0) {
        memset(result->digits, 0, result->capacity * sizeof(int));
        result->size = 0;
        
        // Add z0
        for (int i = 0; i < z0->size; i++) {
            result->digits[i] += z0->digits[i];
        }
        
        // Add z1 shifted by mid
        if (z1) {
            for (int i = 0; i < z1->size; i++) {
                result->digits[i + mid] += z1->digits[i];
            }
        }
        
        // Add z2 shifted by 2*mid
        for (int i = 0; i < z2->size; i++) {
            result->digits[i + 2 * mid] += z2->digits[i];
        }
        
        bigint_normalize(result);
        
        if (z1) {
            bigint_free(z1);
        }
    }
    
    bigint_free(a_low);
    bigint_free(a_high);
    bigint_free(b_low);
    bigint_free(b_high);
    bigint_free(z0);
    bigint_free(z2);
}

int main(int argc, char **argv) {
    MPI_Init(&argc, &argv);
    
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    
    if (argc < 3) {
        if (rank == 0) {
            printf("Usage: %s <num1> <num2> [algorithm]\n", argv[0]);
            printf("  num1, num2: big integers as strings\n");
            printf("  algorithm: 'regular' or 'karatsuba' (default: regular)\n");
        }
        MPI_Finalize();
        return 1;
    }
    
    char *num1_str = argv[1];
    char *num2_str = argv[2];
    char *algorithm = (argc > 3) ? argv[3] : "regular";
    
    BigInt *a = NULL;
    BigInt *b = NULL;
    BigInt *result = NULL;
    
    if (rank == 0) {
        a = bigint_from_string(num1_str);
        b = bigint_from_string(num2_str);
        
        printf("Number 1: %s\n", num1_str);
        printf("Number 2: %s\n", num2_str);
    }
    
    // Broadcast sizes
    int a_size, b_size;
    if (rank == 0) {
        a_size = a->size;
        b_size = b->size;
    }
    MPI_Bcast(&a_size, 1, MPI_INT, 0, MPI_COMM_WORLD);
    MPI_Bcast(&b_size, 1, MPI_INT, 0, MPI_COMM_WORLD);
    
    // Allocate and broadcast BigInts
    if (rank != 0) {
        a = bigint_init(a_size + 10);
        b = bigint_init(b_size + 10);
        a->size = a_size;
        b->size = b_size;
    }
    
    MPI_Bcast(a->digits, a_size, MPI_INT, 0, MPI_COMM_WORLD);
    MPI_Bcast(b->digits, b_size, MPI_INT, 0, MPI_COMM_WORLD);
    
    int result_capacity = a_size + b_size + 100;
    result = bigint_init(result_capacity);
    
    // Time the computation
    double start_time = MPI_Wtime();
    
    if (strcmp(algorithm, "karatsuba") == 0) {
        if (rank == 0) {
            printf("Using Karatsuba algorithm with %d processes\n", size);
        }
        bigint_karatsuba_mpi(a, b, result, rank, size);
    } else {
        if (rank == 0) {
            printf("Using Regular O(n²) algorithm with %d processes\n", size);
        }
        
        BigInt *partial_result = bigint_init(result_capacity);
        bigint_multiply_mpi(a, b, partial_result, rank, size);
        
        // Reduce partial results
        MPI_Reduce(partial_result->digits, result->digits, result_capacity, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
        
        if (rank == 0) {
            bigint_normalize(result);
        }
        
        bigint_free(partial_result);
    }
    
    double end_time = MPI_Wtime();
    double elapsed = end_time - start_time;
    
    if (rank == 0) {
        char *result_str = bigint_to_string(result);
        printf("\nResult: %s\n", result_str);
        printf("Time: %.6f seconds\n", elapsed);
        free(result_str);
    }
    
    bigint_free(a);
    bigint_free(b);
    bigint_free(result);
    
    MPI_Finalize();
    return 0;
}

